core/modules/memman/MemMan.h

```diff
@@ -0,0 +1,177 @@
+
+    enum LockOptions {
+        NOLOCK,               //< Item should not be locked
+        MUSTLOCK,             //< Item must be locked or declare failure
+        FLEXIBLE              //< Item may be locked if possible
+    };
+
+    LockOptions theData;      //< Lock options for the table's data
+    LockOptions theIndex;     //< Lock options for the table's index, if any
+
+    //-----------------------------------------------------------------------------
+    //! Constructor
+    //!
+    //! @param  tabName   is the name of the table.
+    //! @param  optData   lock options for the table's data
+    //! @param  optIndex  lock options for the table's index
+    //-----------------------------------------------------------------------------
+
+    TableInfo(std::string const &tabName,
+              LockOptions optData=MUSTLOCK,
+              LockOptions optIndex=NOLOCK)
+             : tableName(tabName), theData(optData), theIndex(optIndex)
+             {}
+};
+
+//-----------------------------------------------------------------------------
+//! @brief The memory manager.
+//!
+//! The MemMan is an abstract class the defines the interface to the memory
+//! manager that is used to lock database chunks in memory.
+//-----------------------------------------------------------------------------
+
+class MemMan {
+public:
+
+    //-----------------------------------------------------------------------------
+    //! @brief Create a memory manager and initialize for processing.
+    //!
+    //! @param  maxBytes   - Maximum amount of memory that can be used
+    //! @param  dbPath     - Path to directory where the database resides
+    //!
+    //! @return !0: The pointer to the memory manager.
+    //! @return  0: A manager could not be created.
+    //-----------------------------------------------------------------------------
+
+    static MemMan *create(unsigned long long maxBytes,
+                          std::string const &dbPath
+                         );
+
+    //-----------------------------------------------------------------------------
+    //! @brief Lock a set of tables in memory for a particular chunk.
+    //!
+    //! @param  tables - Reference to the tables to process.
+    //! @param  chunk  - The chunk number associated with the tables.
+    //!
+    //! @return =0     - Nothing was locked. The errno variable holds the
+    //!                  reason, as follows:
+    //!                  EFAULT - filesystem or memory error
+    //!                  ENOENT - a chunk was missing
+    //!                  ENOMEM - insufficient memory to fully satisfy request
+    //! @return !0     - Is the resource handle associated with this request.
+    //-----------------------------------------------------------------------------
+
+    typedef unsigned int Handle;
+
+    virtual Handle lock(std::vector<TableInfo> const& tables,
+                        unsigned int chunk
```

Comment:
Yes, I would prefer an explicit typedef for chunk ID. I used unsigned because I don't want to deal with negative numbers. I can change it back to int until we get an official typedef.